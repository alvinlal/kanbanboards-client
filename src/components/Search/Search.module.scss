.search {
  width: 496px;
  height: auto;
  position: relative;

  .search__bar {
    width: 100%;
    height: 48px;
    border-radius: 5px;
    background-color: white;
    display: flex;
    justify-content: space-between;
    align-items: center;
    padding: 0px 16px;

    .search__input {
      display: flex;
      width: calc(100% - 24px);
      align-items: center;
      svg {
        color: $text-200;
      }

      input[type='search'] {
        width: 100%;
        height: 100%;
        margin-left: 10px;
        color: $text-300;
        border: none;
        @include text-base;
      }
    }
  }

  .search__results {
    max-height: 700px;
    overflow-y: auto;
    height: auto;
    width: 100%;
    padding: 16px 0px;
    border-radius: 5px;
    box-shadow: $sh-allsides-large;
    background-color: white;
    position: absolute;
    top: 80px;

    a:hover {
      background-color: $grey-300;
    }

    .search__result {
      width: 100%;
      height: 56px;
      color: $text-300;
      padding: 8px 28px;
      display: flex;
      justify-content: space-between;
      align-items: center;
      cursor: pointer;
      font-weight: $semiBold;
      text-decoration: none;

      h6 {
        white-space: nowrap;
        overflow: hidden;
        text-overflow: ellipsis;
      }

      svg {
        color: $text-300;
        margin-left: 16px;
        @media screen and (max-width: $sm) {
          width: 28px;
          height: 28px;
        }
      }
    }
  }

  .search__noresult {
    border: 2px solid $error-200;
    h6 {
      color: $error-200;
    }
  }

  @media screen and (max-width: $sm) {
    width: 100%;

    .search__bar {
      height: 42px;
    }
    .spinner {
      width: 20px;
      height: 20px;
    }
  }
}
