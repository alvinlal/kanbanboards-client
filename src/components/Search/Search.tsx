import {
  ChevronRightIcon,
  MagnifyingGlassIcon,
} from '@heroicons/react/24/outline';
import React, { useState, useCallback, useRef } from 'react';
import throttle from 'lodash.throttle';
import { Link } from 'react-router-dom';
import SearchResponse from '../../api/search/types/SearchResponse';
import Spinner from '../Spinner/Spinner';
import styles from './Search.module.scss';
import search from '../../api/search';

const Search: React.FC = () => {
  const [results, setResults] = useState<SearchResponse | null>(null);
  const [isResultsVisible, setIsResultsVisible] = useState(false);
  const [loading, setLoading] = useState(false);
  const searchInputRef = useRef<HTMLInputElement>(null);

  const handleChange = useCallback(
    async (e: React.ChangeEvent<HTMLInputElement>) => {
      if (!e.target.value) {
        setResults(null);
        return;
      }
      try {
        setLoading(true);
        const { data } = await search(e.target.value);
        if (searchInputRef.current?.value) {
          setResults(data);
        }
      } catch (err) {
        // send error to monitoring service
      } finally {
        setLoading(false);
      }
    },
    []
  );

  // eslint-disable-next-line react-hooks/exhaustive-deps
  const throttledHandleChange = useCallback(throttle(handleChange, 300), []);

  return (
    <div className={styles.search}>
      <div className={styles.search__bar}>
        <div className={styles.search__input}>
          <MagnifyingGlassIcon width={32} height={32} strokeWidth={3} />
          <input
            type="text"
            placeholder="Search"
            ref={searchInputRef}
            onChange={throttledHandleChange}
            onBlur={(e) => {
              if (e.relatedTarget?.id !== 'search-result') {
                setIsResultsVisible(false);
              }
            }}
            onFocus={() => {
              setIsResultsVisible(true);
            }}
          />
        </div>
        {loading && <Spinner />}
      </div>
      {/* eslint-disable-next-line no-nested-ternary */}
      {results
        ? results.length
          ? isResultsVisible && (
              <div
                className={styles.search__results}
                data-testid="search-result-container"
              >
                {results.map((result) => (
                  <Link
                    key={result._id}
                    className={styles.search__result}
                    to={`/board/${result._id}`}
                    id="search-result"
                  >
                    <h6>{result.title}</h6>
                    <ChevronRightIcon width={32} height={32} />
                  </Link>
                ))}
              </div>
            )
          : isResultsVisible && (
              <div
                className={`${styles.search__results}  ${styles.search__noresult}`}
                data-testid="search-result-container"
              >
                <div className={styles.search__result}>
                  <h6>No Results !</h6>
                </div>
              </div>
            )
        : null}
    </div>
  );
};

export default Search;
