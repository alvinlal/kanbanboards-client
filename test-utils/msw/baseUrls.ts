/* eslint-disable import/prefer-default-export */

export const apiEndPoint = (path: string) => {
  return import.meta.env.VITE_API_ENDPOINT + path;
};
