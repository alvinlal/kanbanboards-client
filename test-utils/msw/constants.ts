export default {
  JWT_SECRET: 'c97a879193249d1ba0d518f6947637e1',
};
