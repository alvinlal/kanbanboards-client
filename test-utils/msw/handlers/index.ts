import authHandlers from './auth';
import boardHandlers from './board';
import searchHandlers from './search';

export default [...authHandlers, ...boardHandlers, ...searchHandlers];
