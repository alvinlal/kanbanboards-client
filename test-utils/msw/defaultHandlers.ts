import { ResponseComposition, RestContext, RestRequest } from 'msw';

export default {
  LOADING: (_: RestRequest, res: ResponseComposition, ctx: RestContext) =>
    res(ctx.delay('infinite')),
  500: (_: RestRequest, res: ResponseComposition, ctx: RestContext) =>
    res(
      ctx.status(500),
      ctx.json({
        statusCode: 500,
        message: 'Internal server error',
      })
    ),
  ERR_NETWORK: (_: RestRequest, res: ResponseComposition) =>
    res.networkError('failed to connect'),
};
