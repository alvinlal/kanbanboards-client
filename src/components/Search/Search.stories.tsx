import { ComponentMeta, ComponentStory } from '@storybook/react';
import { rest } from 'msw';
import { MemoryRouter } from 'react-router-dom';
import { userEvent, within } from '@storybook/testing-library';
import { apiEndPoint } from '../../../test-utils/msw/baseUrls';
import Search from './Search';
import defaultHandlers from '../../../test-utils/msw/defaultHandlers';

export default {
  title: 'Components/Search',
  component: Search,
  decorators: [
    (Story) => (
      <div
        style={{
          backgroundColor: '#4174fa',
          width: '100%',
          height: '88px',
          paddingTop: '20px',
          paddingLeft: '60px',
        }}
      >
        <Story />
      </div>
    ),
  ],
} as ComponentMeta<typeof Search>;

export const Default: ComponentStory<typeof Search> = () => (
  <MemoryRouter>
    <Search />
  </MemoryRouter>
);

export const Loading: ComponentStory<typeof Search> = () => (
  <MemoryRouter>
    <Search />
  </MemoryRouter>
);

Loading.parameters = {
  msw: {
    handlers: [rest.get(apiEndPoint('/search'), defaultHandlers.LOADING)],
  },
};

Loading.play = async ({ canvasElement }) => {
  const canvas = within(canvasElement);

  await userEvent.type(canvas.getByRole('textbox'), 'my project');
};

export const WithResults: ComponentStory<typeof Search> = () => (
  <MemoryRouter>
    <Search />
  </MemoryRouter>
);

WithResults.play = async ({ canvasElement }) => {
  const canvas = within(canvasElement);

  await userEvent.type(canvas.getByRole('textbox'), 'my project');
};

export const NoResults: ComponentStory<typeof Search> = () => (
  <MemoryRouter>
    <Search />
  </MemoryRouter>
);

NoResults.parameters = {
  msw: {
    handlers: [
      rest.get(apiEndPoint('/search'), (_, res, ctx) => {
        return res(ctx.json([]));
      }),
    ],
  },
};

NoResults.play = async ({ canvasElement }) => {
  const canvas = within(canvasElement);

  await userEvent.type(canvas.getByRole('textbox'), 'my project');
};
