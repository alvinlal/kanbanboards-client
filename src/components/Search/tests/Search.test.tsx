/*
  Requirements:-
  1. Should display loading spinner while results are loading
  2. Should hide loading spinner after results have loaded
  3. Should display search results correctly
  4. Should display 'No Results!' when there is no results
  5. Should go to correct url while clicking a search result
  6. Should hide search results on blur
  7. Should show search results on focus
  9. Should match snapshot
  8. Should match api request loading snapshot
  10. Should match api request finished snapshot
*/

import { fireEvent, render, screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { BrowserRouter, Router } from 'react-router-dom';
import { createMemoryHistory } from 'history';
import { faker } from '@faker-js/faker';
import renderer from 'react-test-renderer';
import Search from '../Search';
import { server, rest } from '../../../../test-utils/msw/server';
import { apiEndPoint } from '../../../../test-utils/msw/baseUrls';
import defaultHandlers from '../../../../test-utils/msw/defaultHandlers';

describe('Search.tsx', () => {
  // setup
  const loadingSpinnerTestId = 'spinner';
  const searchResultContainerTestId = `search-result-container`;

  const typeIntoSearch = async (query: string) => {
    const queryInputElement = screen.getByRole('searchbox') as HTMLInputElement;

    await userEvent.type(queryInputElement, query);

    return queryInputElement;
  };

  // tests
  it('Should display loading spinner while results are loading', async () => {
    let loadingSpinner = screen.queryByTestId(loadingSpinnerTestId);
    expect(loadingSpinner).not.toBeInTheDocument();

    server.use(rest.get(apiEndPoint('/search'), defaultHandlers.LOADING));

    render(<Search />, { wrapper: BrowserRouter });

    await typeIntoSearch('my project');

    await waitFor(() => {
      loadingSpinner = screen.getByTestId(loadingSpinnerTestId);
      expect(loadingSpinner).toBeInTheDocument();
    });
  });

  it('Should hide loading spinner after results have loaded', async () => {
    render(<Search />, { wrapper: BrowserRouter });

    await typeIntoSearch('my project');

    await waitFor(() => {
      expect(screen.queryByTestId(loadingSpinnerTestId)).not.toBeInTheDocument();
    });
  });

  it('Should display search results correctly', async () => {
    const results = [
      { title: 'my project 1', board_id: faker.datatype.uuid() },
      { title: 'my project 2', board_id: faker.datatype.uuid() },
      { title: 'my project 3', board_id: faker.datatype.uuid() },
    ];

    server.use(
      rest.get(apiEndPoint('/search'), async (_, res, ctx) => {
        return res(ctx.json(results));
      })
    );

    render(<Search />, { wrapper: BrowserRouter });

    await typeIntoSearch('my project');

    await waitFor(() => {
      const allSearchResults = screen.getAllByRole('link');

      expect(allSearchResults).toHaveLength(results.length);
    });
  });

  it("Should display 'No Results!' when there is no results", async () => {
    server.use(
      rest.get(apiEndPoint('/search'), async (_, res, ctx) => {
        return res(ctx.json([]));
      })
    );

    render(<Search />, { wrapper: BrowserRouter });

    await typeIntoSearch('my project');
    await waitFor(() => {
      const noResultsFound = screen.getByText('No Results !');

      expect(noResultsFound).toBeInTheDocument();
    });
  });

  it('Should go to correct url while clicking a search result', async () => {
    // eslint-disable-next-line @typescript-eslint/naming-convention
    const board_id = faker.datatype.uuid();
    const pathname = `/board/${board_id}`;
    server.use(
      rest.get(apiEndPoint('/search'), async (_, res, ctx) => {
        return res(
          ctx.json([
            {
              title: 'my project',
              board_id,
            },
          ])
        );
      })
    );

    const history = createMemoryHistory();
    history.push = jest.fn();

    render(
      <Router location={history.location} navigator={history}>
        <Search />
      </Router>
    );

    await typeIntoSearch('project 1');

    await userEvent.click(await screen.findByRole('link'));

    expect(history.push).toHaveBeenCalledWith(
      expect.objectContaining({
        pathname,
      }),
      undefined,
      {
        preventScrollReset: undefined,
        relative: undefined,
        replace: false,
        state: undefined,
      }
    );
  });

  it('Should hide search results on blur', async () => {
    render(<Search />, { wrapper: BrowserRouter });

    await typeIntoSearch('my project');

    await waitFor(() => {
      expect(screen.getByTestId(searchResultContainerTestId)).toBeInTheDocument();
    });

    await fireEvent.blur(screen.getByRole('searchbox'));

    await waitFor(() => {
      expect(screen.queryByTestId(searchResultContainerTestId)).not.toBeInTheDocument();
    });
  });

  it('Should show search results on focus', async () => {
    render(<Search />, { wrapper: BrowserRouter });

    await typeIntoSearch('my project');

    await fireEvent.blur(screen.getByRole('searchbox'));

    await waitFor(() => {
      expect(screen.queryByTestId(searchResultContainerTestId)).not.toBeInTheDocument();
    });
    await fireEvent.focus(screen.getByRole('searchbox'));
    await waitFor(() => {
      expect(screen.getByTestId(searchResultContainerTestId)).toBeInTheDocument();
    });
  });

  it('Should match snapshot', () => {
    const tree = renderer.create(<Search />);
    expect(tree).toMatchSnapshot();
  });

  it('Should match api request loading snapshot', async () => {
    server.use(rest.get(apiEndPoint('/search'), defaultHandlers.LOADING));

    const { asFragment } = render(<Search />, {
      wrapper: BrowserRouter,
    });

    await typeIntoSearch('my project');

    await waitFor(() => {
      screen.getByTestId(loadingSpinnerTestId);
    });

    expect(asFragment()).toMatchSnapshot();
  });

  it('Should match api request finished snapshot', async () => {
    const { asFragment } = render(<Search />, {
      wrapper: BrowserRouter,
    });

    await typeIntoSearch('my project');

    await waitFor(() => {
      screen.getAllByRole('link');
    });

    expect(asFragment()).toMatchSnapshot();
  });
});
