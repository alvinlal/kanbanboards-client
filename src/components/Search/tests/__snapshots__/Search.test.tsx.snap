// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search.tsx Should match api request finished snapshot 1`] = `
<DocumentFragment>
  <div
    class="search"
  >
    <div
      class="search__bar"
    >
      <div
        class="search__input"
      >
        <svg
          aria-hidden="true"
          fill="none"
          height="32"
          stroke="currentColor"
          stroke-width="3"
          viewBox="0 0 24 24"
          width="32"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M21 21l-5.197-5.197m0 0A7.5 7.5 0 105.196 5.196a7.5 7.5 0 0010.607 10.607z"
            stroke-linecap="round"
            stroke-linejoin="round"
          />
        </svg>
        <input
          placeholder="Search"
          type="search"
        />
      </div>
    </div>
    <div
      class="search__results"
      data-testid="search-result-container"
    >
      <a
        class="search__result"
        href="/board/ad223casdfdedas392af32223"
        id="search-result"
      >
        <h6>
          my project 1
        </h6>
        <svg
          aria-hidden="true"
          fill="none"
          height="32"
          stroke="currentColor"
          stroke-width="1.5"
          viewBox="0 0 24 24"
          width="32"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M8.25 4.5l7.5 7.5-7.5 7.5"
            stroke-linecap="round"
            stroke-linejoin="round"
          />
        </svg>
      </a>
      <a
        class="search__result"
        href="/board/ad223cdasdged3f392af32223"
        id="search-result"
      >
        <h6>
          my project 2 asdfasd fasdfas dfasdfasd asdsad asd
        </h6>
        <svg
          aria-hidden="true"
          fill="none"
          height="32"
          stroke="currentColor"
          stroke-width="1.5"
          viewBox="0 0 24 24"
          width="32"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M8.25 4.5l7.5 7.5-7.5 7.5"
            stroke-linecap="round"
            stroke-linejoin="round"
          />
        </svg>
      </a>
      <a
        class="search__result"
        href="/board/ad223cdedagsd3f392af32223"
        id="search-result"
      >
        <h6>
          my project 3
        </h6>
        <svg
          aria-hidden="true"
          fill="none"
          height="32"
          stroke="currentColor"
          stroke-width="1.5"
          viewBox="0 0 24 24"
          width="32"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M8.25 4.5l7.5 7.5-7.5 7.5"
            stroke-linecap="round"
            stroke-linejoin="round"
          />
        </svg>
      </a>
      <a
        class="search__result"
        href="/board/ad2ewrw23cded3f392af32223"
        id="search-result"
      >
        <h6>
          my project 4
        </h6>
        <svg
          aria-hidden="true"
          fill="none"
          height="32"
          stroke="currentColor"
          stroke-width="1.5"
          viewBox="0 0 24 24"
          width="32"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M8.25 4.5l7.5 7.5-7.5 7.5"
            stroke-linecap="round"
            stroke-linejoin="round"
          />
        </svg>
      </a>
      <a
        class="search__result"
        href="/board/ad22d3agsdcded3f392af32223"
        id="search-result"
      >
        <h6>
          my project 5
        </h6>
        <svg
          aria-hidden="true"
          fill="none"
          height="32"
          stroke="currentColor"
          stroke-width="1.5"
          viewBox="0 0 24 24"
          width="32"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M8.25 4.5l7.5 7.5-7.5 7.5"
            stroke-linecap="round"
            stroke-linejoin="round"
          />
        </svg>
      </a>
      <a
        class="search__result"
        href="/board/ad2agghreq23cded3f392af32223"
        id="search-result"
      >
        <h6>
          my project 6
        </h6>
        <svg
          aria-hidden="true"
          fill="none"
          height="32"
          stroke="currentColor"
          stroke-width="1.5"
          viewBox="0 0 24 24"
          width="32"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M8.25 4.5l7.5 7.5-7.5 7.5"
            stroke-linecap="round"
            stroke-linejoin="round"
          />
        </svg>
      </a>
      <a
        class="search__result"
        href="/board/ad22kljdrg3cded3f392af32223"
        id="search-result"
      >
        <h6>
          my project 7
        </h6>
        <svg
          aria-hidden="true"
          fill="none"
          height="32"
          stroke="currentColor"
          stroke-width="1.5"
          viewBox="0 0 24 24"
          width="32"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M8.25 4.5l7.5 7.5-7.5 7.5"
            stroke-linecap="round"
            stroke-linejoin="round"
          />
        </svg>
      </a>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Search.tsx Should match api request loading snapshot 1`] = `
<DocumentFragment>
  <div
    class="search"
  >
    <div
      class="search__bar"
    >
      <div
        class="search__input"
      >
        <svg
          aria-hidden="true"
          fill="none"
          height="32"
          stroke="currentColor"
          stroke-width="3"
          viewBox="0 0 24 24"
          width="32"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M21 21l-5.197-5.197m0 0A7.5 7.5 0 105.196 5.196a7.5 7.5 0 0010.607 10.607z"
            stroke-linecap="round"
            stroke-linejoin="round"
          />
        </svg>
        <input
          placeholder="Search"
          type="search"
        />
      </div>
      <div
        class="spinner"
        data-testid="spinner"
      />
    </div>
  </div>
</DocumentFragment>
`;

exports[`Search.tsx Should match snapshot 1`] = `
<div
  className="search"
>
  <div
    className="search__bar"
  >
    <div
      className="search__input"
    >
      <svg
        aria-hidden="true"
        fill="none"
        height={32}
        stroke="currentColor"
        strokeWidth={3}
        viewBox="0 0 24 24"
        width={32}
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M21 21l-5.197-5.197m0 0A7.5 7.5 0 105.196 5.196a7.5 7.5 0 0010.607 10.607z"
          strokeLinecap="round"
          strokeLinejoin="round"
        />
      </svg>
      <input
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        placeholder="Search"
        type="search"
      />
    </div>
  </div>
</div>
`;
